qa/tasks/cephfs/test_acls.py

```diff
@@ -0,0 +1,411 @@
+
+    def test_acl_mask(self):
+        self.setup()
+        acls = [self.u_acl, 'mask::---']
+        rm_acls = [self.rm_u_acl, 'mask::']
+
+        # preliminary test
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        for acl in acls:
+            self.mount_a.run_as_root(['setfacl', '-m', acl, self.testdir1])
+        self.assert_acls_are_('present', acls, self.testdir1)
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+        for acl in rm_acls:
+            self.mount_a.run_as_root(['setfacl', '-x', acl, self.testdir1])
+        self.assert_acls_are_('absent', acls, self.testdir1)
+        self.assert_dirs_are_inaccessible(self.testdir1)
+
+    def test_default_acls(self):
+        self.setup(mkdir=False)
+        acls = [self.du_acl, self.dg_acl, self.do_acl]
+        check_for_acls = [self.du_acl, self.dg_acl, self.u_acl, self.g_acl]
+        rm_acls = [self.rm_du_acl, self.rm_dg_acl, self.rm_u_acl, self.rm_g_acl]
+
+        for acl in acls:
+            self.mount_a.run_as_root(['setfacl', '-m', acl, self.mount_a.mountpoint])
+        self.create_dir_with_appropriate_perms()
+        self.assert_acls_are_('present', acls, self.testdirs)
+        self.assert_dirs_are_accessible(self.testdirs)
+
+        for d in self.testdirs:
+            for acl in rm_acls:
+                self.mount_a.run_as_root(['setfacl', '-x', acl, d])
+        self.assert_acls_are_('absent', check_for_acls, self.testdirs)
+        self.assert_dirs_are_inaccessible(self.testdirs)
+
+    def test_acl_remove_only_default(self):
+        self.setup(mkdir=False)
+
+        for acl in self.default_acls:
+            self.mount_a.run_as_root(['setfacl', '-m', acl, self.mount_a.mountpoint])
+        self.create_dir_with_appropriate_perms()
+        self.assert_acls_are_('present', self.default_acls, self.testdirs)
+        self.assert_dirs_are_accessible(self.testdir1)
+
+        self.mount_a.run_as_root(['setfacl', '-k', self.testdir1])
+        self.assert_acls_are_('absent', self.default_acls, self.testdir1)
+        # since non default ACLs would left over files.
+        self.assert_dirs_are_accessible(self.testdirs)
+
+    def test_acl_remove_all_acls(self):
+        self.setup(mkdir=False)
+
+        for acl in self.all_acls:
+            self.mount_a.run_as_root(['setfacl', '-m', acl, self.mount_a.mountpoint])
+        self.create_dir_with_appropriate_perms()
+        self.assert_acls_are_('present', self.all_acls, self.testdirs)
+        self.assert_dirs_are_accessible(self.testdirs)
+
+        for d in self.testdirs:
+            self.mount_a.run_as_root('setfacl -b ' + d)
+        self.assert_acls_are_('absent', self.default_acls, self.testdirs)
+        self.assert_dirs_are_inaccessible(self.testdirs)
+
+    #######################################################
+    #
+    # Test recursive cases
+    #
+    ####################################################
+
+    def test_options_Rx_for_all_acls(self):
```

Comment:
I think the "recursive" tests here are really just testing the functionality of setfacl. setfacl will just walk the tree recursively and update the xattr of each object. I think you can probably drop most of these recursive tests.

Reply:
> I think the "recursive" tests here are really just testing the functionality of setfacl. setfacl will just walk the tree recursively and update the xattr of each object. I think you can probably drop most of these recursive tests.

I admit that I did feel this after I completed writing these tests. Originally, this was a single testcase (named `set, view and remove ACLs recursively`) among the rest, so I couldn't see that straightaway.

Yesterday, I had a thought on expanding the recursive tests for default ACLs. Right now, the testcase for default ACLs has the following steps:

1. Apply ACLs to the mountpoint.
2. Create directories inside
3. Check if default ACLs worked as expected on newly created directories.

Following could be a better case, if we want it to keep it in the first place -

1. Create directories
2. Apply ACLs to the mountpoint.
3. Check if default ACLs are on mountpoint and on the directories within it.
4. Create directories within directories
5. Check if default ACLs corresponding to default ACLs are present on newly created directories.

Is this worth doing/keeping? Or... this is just a complex version of [test_default_acls()](https://github.com/ceph/ceph/pull/26477/files#diff-c2417cd8e2afc1b19952425d5f25534fR254)? In case we want to keep these tests, the current approach of recursive tests isn't good enough, IMO.

> FWIW, I did ask around and didn't find any comprehensive testsuites for this. xfstests has some ACL related tests, and I saw a fork of pjdfstests that had some too. Maybe we should just look at leveraging those for this testing? Is that unreasonable?

I can't say that myself since I wasn't aware of these tests until now.
 
> In any case, I'm a little unclear on what the exact requirements are here. Are you looking to specifically test setting/removing ACLs? Or are you more interested in testing whether ACLs are being properly enforced for different POSIX filesystem operations (of which the setxattr that handles the ACL change is only one).

Actually both. When I started with this PR, I had the intention to write tests that would make sure that ACLs are working as one would expect them to work on ext4 no matter what underlying FS is used with CephFS.

Reply:
Yes, we'd definitely want to test ACL inheritance. That said, will testing beyond one level of inheritance be useful?

If we're interested in ensuring ACL compliance over kernel-based filesystems, might we be better served by adapting existing testsuites? I imagine they will be more comprehensive than anything we could come up with.

FWIW:

LTP has this: https://github.com/linux-test-project/ltp/tree/master/testcases/kernel/fs/acl

...pretty old, but would be reasonably easy to adapt to cephfs.

I looked closer at xfstests: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git

If you do a git grep in there for "setfacl", there are some tests that call it. Most of them seem to be regression tests for specific bugs though, so I'm not sure if they're really quite what we'd want.

This guy has a small suite of ACL tests that might be useful: https://github.com/Hnasar/pjdfstest/tree/master/tests/xacl

...but it's not clear to me why he didn't contribute them back to upstream pjdfstests though. They seem to be BSD licensed, so we should be able to use them for this if we wish (just be sure to attribute them properly).

Reply:
> Yes, we'd definitely want to test ACL inheritance. That said, will testing beyond one level of inheritance be useful?
>

Hmm, I get it. The new steps only adds complexity but the testcase remains the same. I'll go ahead and delete recursive test cases.

> If we're interested in ensuring ACL compliance over kernel-based filesystems, might we be better served by adapting existing testsuites? I imagine they will be more comprehensive than anything we could come up with.
>
> FWIW:
> 
> LTP has this: https://github.com/linux-test-project/ltp/tree/master/testcases/kernel/fs/acl
> 
> ...pretty old, but would be reasonably easy to adapt to cephfs.
> 
> I looked closer at xfstests: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git
> 
> If you do a git grep in there for "setfacl", there are some tests that call it. Most of them seem to be regression tests for specific bugs though, so I'm not sure if they're really quite what we'd want.
> 
> This guy has a small suite of ACL tests that might be useful: https://github.com/Hnasar/pjdfstest/tree/master/tests/xacl
> 
> ...but it's not clear to me why he didn't contribute them back to upstream pjdfstests though. They seem to be BSD licensed, so we should be able to use them for this if we wish (just be sure to attribute them properly).

Ok. I'll take a look and create a list of testcases these testsuites cover so that we can compare them with ours. Do we want to translate these test cases to Python or would be prefer to use them as it is?

Reply:
> Ok. I'll take a look and create a list of testcases these testsuites cover so that we can compare them with ours. Do we want to translate these test cases to Python or would be prefer to use them as it is?

I think you get to make that call, since you're doing the work. :) It really comes down to what you think will be easiest to do and (more importantly) maintain over the long haul.

Reply:
I think Python's better from maintainability POV. Especially after the rewriting I did late last week and early this week, the code in test methods is nearly same as writing English.

@jtlayton @batrick @ukernel @vshankar 
In case there is a different opinion on this, please let me know.